ompi/mca/coll/han/coll_han_subcomms.c

```diff
@@ -0,0 +1,463 @@
+ *                         All rights reserved.
+ * Copyright (c) 2014-2015 Research Organization for Information Science
+ *                         and Technology (RIST). All rights reserved.
+ * Copyright (c) 2015      Intel, Inc. All rights reserved.
+ * Copyright (c) 2020      Bull S.A.S. All rights reserved.
+ *
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+/**
+ * @file
+ *
+ * Warning: this is not for the faint of heart -- don't even bother
+ * reading this source code if you don't have a strong understanding
+ * of nested data structures and pointer math (remember that
+ * associativity and order of C operations is *critical* in terms of
+ * pointer math!).
+ */
+
+#include "ompi_config.h"
+
+#include "mpi.h"
+#include "coll_han.h"
+#include "coll_han_dynamic.h"
+
+
+/*
+ * Local functions
+ */
+static void create_intranode_comm_new(ompi_communicator_t *,
+                                      ompi_communicator_t **);
+static void create_internode_comm_new(ompi_communicator_t *,
+                                      int, int,
+                                      ompi_communicator_t **);
+static void create_intranode_comm(ompi_communicator_t *,
+                                  const char *,
+                                  int,
+                                  ompi_communicator_t **);
+static void create_internode_comm(ompi_communicator_t *,
+                                  const char *,
+                                  int, int,
+                                  ompi_communicator_t **);
+
+/**
+ * Create a sub-communicator containing the ranks that share my node.
+ *
+ * @param comm (IN)          original communicator for the collective
+ *                           target module priority
+ * @param sub_comm (OUT)     created sub-communicator
+ */
+static void create_intranode_comm_new(ompi_communicator_t *comm,
+                                  ompi_communicator_t **sub_comm)
+{
+    ompi_comm_split_type(comm, MPI_COMM_TYPE_SHARED, 0,
+                         (opal_info_t *)(&ompi_mpi_info_null), sub_comm);
+    return;
+}
+
+/**
+ * Create a sub-communicator containing one rank per node.
+ *
+ * @param comm (IN)          original communicator for the collective
+ * @param my_rank (IN)       my rank in comm
+ * @param intra_rank (IN)    local rank in the intra-node sub-communicator
+ * @param sub_comm (OUT)     created sub-communicator
+ */
+static void create_internode_comm_new(ompi_communicator_t *comm,
+                                  int my_rank,
+                                  int intra_rank,
+                                  ompi_communicator_t **sub_comm)
+{
+    ompi_comm_split(comm, intra_rank, my_rank, sub_comm, false);
+    return;
+}
+
+/*
+ * Routine that creates the local hierarchical sub-communicators
+ * Called each time a collective is called.
+ * comm: input communicator of the collective
+ */
+void mca_coll_han_comm_create_new(struct ompi_communicator_t *comm,
```

Comment:
In the current OMPI this entire routine can be done mostly locally. The local ranking is known from PMIx as well as the daemon/node where each process is places. What we don't know is who will be the remote leader, and this might be the only place we need to rely on another collective.

For more info take a look at the email archive of OMPI-devel for the thread with subject "Taking advantage of PMIx: Hierarchical collective support"

Reply:
I'll take this comment back, at least partially. In order to take advantage of the PMIx info we would need to be able to reserve 2 comm_id by some other means, and unfortunately this is not possible today in OMPI. But we should think about it ;)